src/v/cluster/cluster_epoch_generator.cc

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/cluster_epoch_generator.h"
+
+#include "base/vassert.h"
+#include "cluster/controller_stm.h"
+#include "cluster/logger.h"
+
+#include <seastar/core/sleep.hh>
+#include <seastar/core/smp.hh>
+
+namespace cluster {
+
+namespace {
+
+using namespace std::chrono_literals;
+
+// TODO(cloud-topics): make these configuration knobs
+constexpr ss::lowres_clock::duration epoch_cache_timeout = 1s;
+// The maximum amount of time to tolerate the same epoch being returned
+constexpr ss::lowres_clock::duration max_same_epoch_duration = 1min;
+
+// NOLINTNEXTLINE(*-avoid-reference-coroutine-parameters)
+ss::future<> force_increment_loop(controller_stm& stm, ss::abort_source& as) {
+    model::offset last_offset = co_await stm.bootstrap_committed_offset();
+    while (!as.abort_requested()) {
+        try {
+            co_await ss::sleep_abortable<ss::lowres_clock>(
+              max_same_epoch_duration, as);
+        } catch (const ss::sleep_aborted&) {
+            co_return;
+        }
+        model::offset new_offset = co_await stm.bootstrap_committed_offset();
+        if (new_offset > last_offset) {
+            last_offset = new_offset;
+            continue;
+        }
+        auto result = co_await stm.quorum_write_empty_batch(
```

Comment:
would some of these co_await stm's ever throw? might make sense to wrap the entire increment loop in a try block so the loop will auto restart?

Reply:
this seems to have been completely rewritten in later commits